File: frontend/src/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import {
  RegisterUserDto,
  LoginUserDto,
  EmailDto,
  UpdateUserDto,
} from 'src/dtos';
import { Observable } from 'rxjs';
import { Router } from '@angular/router';
import { JwtHelperService } from '@auth0/angular-jwt';
import { STRING_TYPE } from '@angular/compiler';

import { UserSearchDto } from '../dtos/req/userSearch.dto';
import { UserDetailDto } from '../dtos/req/userDetail.dto';
import { UserAdminEditDto } from '../dtos/req/userAdminEdit.dto';

@Injectable({ providedIn: 'root' })
export class UserService {
  constructor(
    private http: HttpClient,
    private router: Router,
    private jwtHelper: JwtHelperService
  ) {}

  /**
   * Register a new user
   * @param newUser The user to register
   * @returns The user's token
   */
  public register(newUser: RegisterUserDto) {
    return this.http.post('auth/register', newUser, { responseType: 'text' });
  }

  /**
   * Login an existing user
   * @param user The user to login
   * @returns The user's token
   */
  public login(user: LoginUserDto) {
    return this.http.post('auth/login', user, { responseType: 'text' });
  }

  /**
   * Change data of an existing user
   * @param user the new user data
   * @param id the id of the user to be changed
   * @returns the user's token
   */
  public update(user: UpdateUserDto, id: number) {
    return this.http.put('users/' + id, user, {
      responseType: 'text',
    });
  }

  /**
   * Logout the current user and redirect to the login page
   * @returns void
   */
  public logout() {
    localStorage.removeItem('token');
    this.router.navigate(['/login']);
  }

  /**
   * Reset the password of an existing user
   * @param email The email to the password to reset
   * @returns nothing
   */
  public resetPassword(email: EmailDto) {
    return this.http.put('users/resetpassword', email, {
      responseType: 'text',
    });
  }

  /**
   * Checks if email links to an existing user
   * @param email The email to check for
   * @returns the email
   */
  public checkEmail(email: EmailDto) {
    const params = new HttpParams().set('email', email.email);
    return this.http.get('users/checkemail', { params });
  }

  /**
   * Deletes an existing user in the system.
   *
   * @param id the id of the account that should be deleted
   * @return nothing
   */
  public delete(id: number) {
    return this.http.delete('users/' + id);
  }

  /**
   * Get a list of users with the given search parameters.
   * @param searchParameters: parameters that can be combined as required
   * @return observable list of found users.
   */
  public search(searchParameters: UserSearchDto): Observable<UserDetailDto[]> {
    let params = new HttpParams();

    if (searchParameters.id != null) {
      params = params.set('id', searchParameters.id);
    }

    if (searchParameters.email != null) {
      params = params.set('email', searchParameters.email);
    }

    if (searchParameters.role != null) {
      params = params.set('role', searchParameters.role);
    }

    if (searchParameters.active != null) {
      params = params.set('active', searchParameters.active);
    }

    return this.http.get<UserDetailDto[]>('users', { params }).pipe();
  }

  /**
   * Get the Observable of the user with the specified ID.
   *
   * @param id of the user to get
   * @return the Observable of the user with the specified ID.
   */
  public getById(id: number): Observable<UserDetailDto> {
    return this.http.get<UserDetailDto>(`users/${id}`).pipe();
  }

  /**
   * Update the role and active status for the user with the specified id.
   *
   * @param id of the user to update
   * @param userDto contains the attributes to update
   * @return the Observable for the updated user.
   */
  public updateByAdmin(
    id: number,
    userDto: UserAdminEditDto
  ): Observable<UserDetailDto> {
    return this.http.patch<UserDetailDto>(`users/${id}`, userDto).pipe();
  }
}

File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/UserService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.dtos.req.UserAdminEditDto;
import at.ac.tuwien.sepm.groupphase.backend.dtos.req.UserCredentialsDto;
import at.ac.tuwien.sepm.groupphase.backend.dtos.req.UserDetailDto;
import at.ac.tuwien.sepm.groupphase.backend.dtos.req.UserLoginDto;
import at.ac.tuwien.sepm.groupphase.backend.dtos.req.UserProfileDto;
import at.ac.tuwien.sepm.groupphase.backend.dtos.req.UserRegisterDto;
import at.ac.tuwien.sepm.groupphase.backend.dtos.req.UserResetPasswordDto;
import at.ac.tuwien.sepm.groupphase.backend.dtos.req.UserSearchDto;
import at.ac.tuwien.sepm.groupphase.backend.dtos.req.UserUpdateRequestDto;
import at.ac.tuwien.sepm.groupphase.backend.entity.User;
import at.ac.tuwien.sepm.groupphase.backend.exception.NotFoundException;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;

import java.util.stream.Stream;

public interface UserService extends UserDetailsService {

    /**
     * Find a user in the context of Spring Security based on the username
     * <br>
     * For more information have a look at this tutorial:
     * https://www.baeldung.com/spring-security-authentication-with-a-database
     *
     * @param username the username
     * @return a Spring Security user
     * @throws UsernameNotFoundException is thrown if the specified user does not exist
     */
    @Override
    UserCredentialsDto loadUserByUsername(String username) throws UsernameNotFoundException;

    /**
     * Find an application user based on the username.
     *
     * @param username the username
     * @return an application user
     * @throws NotFoundException if user by username does not exist
     */
    User findApplicationUserByUsername(String username) throws NotFoundException;

    /**
     * Find an application user based on the email address.
     *
     * @param email the email address
     * @return an application user
     * @throws NotFoundException if user by email does not exist
     */
    User findApplicationUserByEmail(String email) throws NotFoundException;

    /**
     * Log in a user.
     *
     * @param userLoginDto login credentials
     * @return the JWT, if successful
     * @throws BadCredentialsException if credentials are bad
     */
    String login(UserLoginDto userLoginDto) throws BadCredentialsException;

    /**
     * Register a new user.
     *
     * @param userRegisterDto registration data
     * @return the JWT, if successful
     */
    String register(UserRegisterDto userRegisterDto);

    /**
     * Updates an already existing user.
     * Checks if the request sender is not trying to change someone else's data
     *
     * @param userUpdateRequestDto the new user data
     * @throws AccessDeniedException in case the sender of the request is trying to change data of another user
     * @throws BadCredentialsException if password is false
     */
    void updateUser(UserUpdateRequestDto userUpdateRequestDto) throws AccessDeniedException, BadCredentialsException;

    /**
     * Reset the password for a given email.
     *
     * @param email emailaddress
     * @throws UsernameNotFoundException if user doesnt exist
     */
    void resetPassword(UserResetPasswordDto email) throws UsernameNotFoundException;

    /**
     * Deletes an account.
     *
     * @param id account Id
     */
    void deleteUser(Long id);

    /**
     * Search for users matching the criteria in {@code searchParameters}.
     * <p>
     * A user is considered matched,
     * if its id matches {@code searchParameters.id},
     * if its email contains {@code searchParameters.email},
     * if its role matches {@code searchParameters.role}.
     * All parameters are optional.
     * </p>
     *
     * @param searchParameters object containing the search parameters to match
     * @return a stream containing users matching the criteria in {@code searchParameters}
     */
    Stream<UserDetailDto> search(UserSearchDto searchParameters);

    /**
     * Get UserDetailDto by id.
     *
     * @param id of the user to be retrieved
     * @return a UserDetailDto of the matching user.
     * @throws NotFoundException if no user with the given ID exists.
     */
    UserDetailDto getById(Long id) throws NotFoundException;

    /**
     * Update user by id with the UserDetailDto provided.
     *
     * @param id of the user to be updated
     * @param userDto attributes to update (only {@code userDetail.role} and {@code userDetail.active}
     * @return the new UserDetailDto of the user
     * @throws NotFoundException if no user with the given ID exists.
     */
    UserDetailDto updateByAdmin(Long id, UserAdminEditDto userDto) throws NotFoundException;

    /**
     * Get UserProfileDto by user id.
     *
     * @param id of the user.
     * @return UserProfileDto of the specified user.
     * @throws NotFoundException if no user with the given ID exists.
     */
    UserProfileDto getProfileById(Long id) throws NotFoundException;
}
